samples/kinesis_file_gstreamer_sample_app.cpp

```diff
@@ -0,0 +1,797 @@
+
+    // key:     trackId
+    // value:   whether application has received the first frame for trackId.
+    map<int, bool> stream_started;
+
+    // stores any error status code reported by StreamErrorCallback.
+    atomic_uint stream_status;
+
+    // When uploading file, store the pts of frames that has flag FRAME_FLAG_KEY_FRAME. When the entire file has been uploaded,
+    // key_frame_pts contains the timetamp of the last fragment in the file. key_frame_pts is then stored into last_fragment_ts
+    // of the file.
+    // Unit: ns
+    uint64_t key_frame_pts;
+
+    // Assuming frame timestamp are relative. Add producer_start_time to each frame's
+    // timestamp to convert them to absolute timestamp. This way fragments dont overlap after token rotation when doing
+    // file uploading.
+    uint64_t producer_start_time;
+
+    unique_ptr<Credentials> credential;
+} CustomData;
+
+namespace com { namespace amazonaws { namespace kinesis { namespace video {
+
+class SampleClientCallbackProvider : public ClientCallbackProvider {
+public:
+
+    UINT64 getCallbackCustomData() override {
+        return reinterpret_cast<UINT64> (this);
+    }
+
+    StorageOverflowPressureFunc getStorageOverflowPressureCallback() override {
+        return storageOverflowPressure;
+    }
+
+    static STATUS storageOverflowPressure(UINT64 custom_handle, UINT64 remaining_bytes);
+};
+
+class SampleStreamCallbackProvider : public StreamCallbackProvider {
+    UINT64 custom_data_;
+public:
+    SampleStreamCallbackProvider(UINT64 custom_data) : custom_data_(custom_data) {}
+
+    UINT64 getCallbackCustomData() override {
+        return custom_data_;
+    }
+
+    StreamConnectionStaleFunc getStreamConnectionStaleCallback() override {
+        return streamConnectionStaleHandler;
+    };
+
+    StreamErrorReportFunc getStreamErrorReportCallback() override {
+        return streamErrorReportHandler;
+    };
+
+    DroppedFrameReportFunc getDroppedFrameReportCallback() override {
+        return droppedFrameReportHandler;
+    };
+
+    FragmentAckReceivedFunc getFragmentAckReceivedCallback() override {
+        return fragmentAckReceivedHandler;
+    };
+
+private:
+    static STATUS
+    streamConnectionStaleHandler(UINT64 custom_data, STREAM_HANDLE stream_handle,
+                                 UINT64 last_buffering_ack);
+
+    static STATUS
+    streamErrorReportHandler(UINT64 custom_data, STREAM_HANDLE stream_handle, UPLOAD_HANDLE upload_handle, UINT64 errored_timecode,
+                             STATUS status_code);
+
+    static STATUS
+    droppedFrameReportHandler(UINT64 custom_data, STREAM_HANDLE stream_handle,
+                              UINT64 dropped_frame_timecode);
+
+    static STATUS
+    fragmentAckReceivedHandler( UINT64 custom_data, STREAM_HANDLE stream_handle,
+                                UPLOAD_HANDLE upload_handle, PFragmentAck pFragmentAck);
+};
+
+class SampleCredentialProvider : public StaticCredentialProvider {
```

Comment:
Not sure if this really adds any value. Perhaps we should use StaticCredentialProvider directly. I think this was an attempt to demonstrate how a custom credential provider might work. However, once these are outsourced into the Common/ directory we can think of adding it back or something. Let me know what you think

Reply:
Yea, I think this a good idea, I think users don't need to be burdened to read this part since this is not the main focus.